Render loop

grokys opened the pull request:
This PR adds a render loop. It's intended to be reasonably non-intrusive, and the beginning of a larger overhaul of our rendering.
- It adds the start of a `RenderTest` sample which we can use to monitor performance/diagnose drawing errors etc
- Adds an `IRenderLoop` interface, a default `DefaultRenderLoop` implementation and a `Win32.RenderLoop` implementation that uses multimedia timers (the default Win32 timer only has the resolution to go up to about 30 fps when requested 60fps)
- It replaces `IRenderRoot.RenderQueueManager` with `IRenderRoot.Renderer` which is of type `IRenderer`
- `IRenderer` will keep track of dirty rectangles (currently unused) and carry out rendering. For rendering it currently just calls the existing rendering infrastructure
- It adds `IRenderRoot.CreateRenderTarget` to allow a render root to create a render target (this currently just calls `IPlatformRenderInterface.CreateRenderer`) as before, but it is intended to decouple the renderer from knowledge of platform handles (which are only defined higher up in the stack)

grokys committed:
Added a render test sample.

To measure FPS with animations.

grokys committed:
WM_PAINT must always be handled.

Timers won't be called if not as WM_PAINT has a higer priority.

grokys committed:
Added a render loop.

Rather than repainting whenever, this adds a render loop which is called
N times per second. IRenderQueueManager and ITopLevelRenderer have been
removed, replaced by IRenderer.

grokys committed:
Make Avalonia.Android compile again.

No idea if it actually works because android doesn't seem to work at all
at the moment.

grokys committed:
Fix FPS counter.

It was updating on every tick rather than every second.

grokys committed:
Merge branch 'master' into render-loop

 Conflicts:
	src/Android/Avalonia.Android/AndroidPlatform.cs
	src/Avalonia.Controls/TopLevel.cs

grokys committed:
Somehow missed a file from merge.

grokys committed:
Rename RenderLoop -> DefaultRenderLoop.

Make it overridable and register it in the windowing subsystem, so that
windowing subsystems can provide custom implementations.

grokys committed:
Use multimedia timer on Win32 for render loop.

grokys committed:
Don't need reentry guard here.

grokys commented:
cc: @kekekeks

grokys committed:
Fix .sln merge error.

grokys committed:
FIx control leak tests.

kekekeks commented on src/Avalonia.SceneGraph/Rendering/Renderer.cs:
```diff
@@ -0,0 +1,67 @@
+
+        public void Attach(IRenderRoot root, IRenderLoop renderLoop)
+        {
+            Contract.Requires<ArgumentNullException>(root != null);
```
Is renderer instance reusable? If it's not why not to use constructor injection or at least check that `Attach` was already called?

grokys replied:
Nope, there is one renderer per top level control. I use `Attach` because it's created with the service locator and there's no way to pass constructor arguments using that.

kekekeks replied:
You could register `Func<TopLevel, IRenderer>` (or custom delegate to be used as factory method) in locator instead

grokys replied:
Yeah, true. Or even register a `RendererFactory`...

kekekeks commented on src/Avalonia.SceneGraph/Rendering/Renderer.cs:
```diff
@@ -0,0 +1,67 @@
+            {
+                Logging.Logger.Information("Renderer", this, "Render target was corrupted. Exception: {0}", ex);
+                _renderTarget.Dispose();
+                _renderTarget = null;
```
I think we are supposed to retry render pass here

grokys replied:
Well the next render pass will happen in 1/60 of a second, so I figured just wait until then.

kekekeks commented on src/Windows/Avalonia.Win32/RenderLoop.cs:
```diff
@@ -0,0 +1,35 @@
+        {
+            timerDelegate = (id, uMsg, user, dw1, dw2) => tick();
+
+            var handle = UnmanagedMethods.timeSetEvent(
```
Are you sure that it's thread safe? MSDN states that
>The multimedia timer runs in its own thread

grokys replied:
Yep, it runs in its own thread, it's the responsibility of whatever subscribes to the render loop to handle passing the notification to the correct thread - I should document that fact.

At the moment we're still going via WM_PAINT to do the rendering so that will move us onto the right thread; later on we'll probably paint on the render thread to an offscreen bitmap and render sections of that on WM_PAINT as needed.

kekekeks commented on src/Avalonia.Controls/TopLevel.cs:
```diff
@@ -232,6 +235,18 @@ protected bool AutoSizing
         /// <inheritdoc/>
+        IRenderTarget IRenderRoot.CreateRenderTarget()
+        {
+            return _renderInterface.CreateRenderer(PlatformImpl.Handle);
```
May be we should move `CreateRenderer` to `ITopLevelImpl`. Windowing platform has more information for negotiations with rendering platform than we do in `TopLevel` class

kekekeks replied:
I think we might even make `ITopLevelImpl` to actually implement `IRenderTarget` by itself. There are some cases (e. g. GTK embedding) when we are trying to obtain render target when native window haven't been created yet. I think we should ask for `DrawingContext` from `ITopLevelImpl` directly and allow it to return `null` if `DrawingContext` can't be created and skip render pass in this case. It will also allow to skip frames when window isn't effectively visible to the user (minimized, out of screen borders, etc).

grokys replied:
Right, that's what I was going for with this. `IRenderRoot` is implemented by `TopLevel` and that can choose how to handle it; including passing the request to `ITopLevelImpl`. 

If I understand you correctly you're talking about losing `IRenderTarget` too, and just having something like `IRenderRoot.CreateDrawingContext` (which TopLevel will choose to pass to `ITopLevelImpl.CreateRenderTarget`)? If so, I think that's a good idea.

kekekeks replied:
Yep, something like that. IRenderTarget isn't that useful right now anyway, since it provides just one method.

kekekeks replied:
Well, `IRenderTarget` *is* useful for tracking state from the rendering system (direct2d/opengl context, offscreen surface, etc), but lifetime management of said render target should be the job of `ITopLevelImpl`, I think.

kekekeks replied:
I think `ITopLevelImpl` implementation should obtain `IRenderTarget` from rendering subsystem (using IPlatformHandle *or* bitmap-backed) and manage it's lifetime by itself.

grokys replied:
Ok, sounds good - though I think that should be a separate PR; I'd like to keep this focused on the render loop change. As I said it's intended to be part of a much larger change. As such does it look good to merge now to you? If so could you merge? I'd like to use the system where the reviewer merges the PR rather than the author.

grokys committed:
Fixed name of method.

It's creating a render target, not a renderer.

grokys committed:
Add IRendererFactory.

I don't like to add factories, but it's better than the previous
`Attach` way of doing things; it's necessary because it's not possible
to pass constuctor parameters using `AvaloniaLocator`.

grokys committed:
Added a comment about RenderLoop threading.

kekekeks commented on src/Avalonia.SceneGraph/Rendering/Renderer.cs:
```diff
@@ -0,0 +1,67 @@
+        {
+            if (_dirty)
+            {
+                _root.Invalidate(new Rect(_root.ClientSize));
```
That's being called on `IRenderLoop` thread, so it might not be safe to call `IRenderRoot` methods here. For now on win32 platform `Invalidate` just calls `ITopLevel::Invalidate` which in turn calls `SendMessage`, so you've got away with that, but it is not safe in general case.

I also propose to add some suffix for events that might trigger on non-ui thread (e. g. `Tick` should be named `TickThreadUnsafe`) to make it easier to track such things.

grokys replied:
Yep, this is done with the knowledge that we go via the message loop for paints. It will be replaced by the real render code in upcoming patches.

grokys replied:
I guess synchronizing with the UI thread here wouldn't be a problem though.

grokys commented:
@kekekeks given that the remaining comments you raised will be fixed in upcoming PRs, are we OK to merge this?

kekekeks committed:
Merge branch 'master' into render-loop
